src/Common/src/Interop/Unix/System.Net.Security.Native/Interop.GssBuffer.cs

```diff
@@ -0,0 +1,64 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class NetSecurityNative
+    {
+        [StructLayout(LayoutKind.Sequential)]
+        internal unsafe struct GssBuffer : IDisposable
+        {
+            internal UInt64 length;
+            internal IntPtr data;
+
+            internal int Copy(byte[] destination, int offset)
+            {
+                Debug.Assert(destination != null, "target destination cannot be null");
+                Debug.Assert(offset >= 0 && offset < destination.Length, "invalid offset " + offset);
+
+                if (data == IntPtr.Zero || length == 0)
+                {
+                    return 0;
+                }
+
+                int bufferLength = Convert.ToInt32(length);
+                int available = destination.Length - offset;  // amount of space in the given buffer
+                if (bufferLength > available)
+                {
+                    throw new NetSecurityNative.GssApiException(SR.Format(SR.net_context_buffer_too_small, bufferLength, available));
+                }
+
+                Marshal.Copy(data, destination, offset, bufferLength);
+                return bufferLength;
+            }
+
+            internal byte[] ToByteArray()
+            {
+                if (data == IntPtr.Zero || length == 0)
+                {
+                    return new byte[0];
+                }
+
+                int bufferLength = Convert.ToInt32(length);
+                byte[] destination = new byte[bufferLength];
+                Marshal.Copy(data, destination, 0, bufferLength);
+                return destination;
+            }
+
+            public void Dispose()
+            {
+                if (data != IntPtr.Zero)
+                {
+                    Interop.NetSecurityNative.ReleaseGssBuffer(data, length);
```

Comment:
> Set data to IntPtr.Zero here so that calling Dispose multiple times is still safe.

Ok. I thought we are not supposed to mutate the field of a struct.